File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_VisualBasic/SolverGenerator/SolverOptions.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: SolverOptions.vb
'
'  Description: Options for the Sudoku solver.
' 
'--------------------------------------------------------------------------

Imports Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques

Namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku
	''' <summary>Options for configuring how the solver does its processing.</summary>
	Public NotInheritable Class SolverOptions
		Implements ICloneable
		''' <summary>The maximum number of solutions the solver should find.</summary>
		Private _maximumSolutionsToFind? As UInteger = 1

		''' <summary>Whether to allow brute-force techniques in the solver.</summary>
		Private _allowBruteForce As Boolean = True

		''' <summary>The techniques to use while solving.</summary>
		Private _techniques As List(Of EliminationTechnique)

		''' <summary>Initializes the SolverOptions.</summary>
		Public Sub New()
		End Sub

		''' <summary>Gets or sets the maximum number of solutions the solver should find.</summary>
		Public Property MaximumSolutionsToFind() As UInteger?
			Get
				Return _maximumSolutionsToFind
			End Get
			Set(ByVal value? As UInteger)
				If value <= 0UI Then
					Throw New ArgumentOutOfRangeException("value")
				End If
				_maximumSolutionsToFind = value
			End Set
		End Property

		''' <summary>Gets or sets the techniques used for solving.</summary>
		Public Property EliminationTechniques() As List(Of EliminationTechnique)
			Get
				If _techniques Is Nothing Then
					_techniques = New List(Of EliminationTechnique)()
				End If
				Return _techniques
			End Get
			Set(ByVal value As List(Of EliminationTechnique))
				_techniques = value
			End Set
		End Property

		''' <summary>Gets or sets whether to allow brute-force techniques in the solver.</summary>
		Public Property AllowBruteForce() As Boolean
			Get
				Return _allowBruteForce
			End Get
			Set(ByVal value As Boolean)
				_allowBruteForce = value
			End Set
		End Property

		''' <summary>Clones the solver options.</summary>
		''' <returns>A clone of the options.</returns>
		Public Function Clone() As SolverOptions
            Dim options As New SolverOptions()
            With options
                .AllowBruteForce = AllowBruteForce
                .MaximumSolutionsToFind = MaximumSolutionsToFind
                .EliminationTechniques = New List(Of EliminationTechnique)()

            End With
            For Each technique In EliminationTechniques
                options.EliminationTechniques.Add(technique.Clone())
            Next technique
            Return options
        End Function

		''' <summary>Clones the solver options.</summary>
		''' <returns>A clone of the options.</returns>
		Private Function ICloneable_Clone() As Object Implements ICloneable.Clone
			Return Clone()
		End Function
	End Class
End Namespace

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_CSharp/SolverGenerator/SolverOptions.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: SolverOptions.cs
//
//  Description: Options for the Sudoku solver.
// 
//--------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques;

namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku
{
	/// <summary>Options for configuring how the solver does its processing.</summary>
	public sealed class SolverOptions : ICloneable
	{
		/// <summary>The maximum number of solutions the solver should find.</summary>
		private uint? _maximumSolutionsToFind = 1;

		/// <summary>Whether to allow brute-force techniques in the solver.</summary>
		private bool _allowBruteForce = true;

		/// <summary>The techniques to use while solving.</summary>
		private List<EliminationTechnique> _techniques;

		/// <summary>Initializes the SolverOptions.</summary>
		public SolverOptions() {}

		/// <summary>Gets or sets the maximum number of solutions the solver should find.</summary>
		public uint? MaximumSolutionsToFind
		{
			get { return _maximumSolutionsToFind; }
			set 
			{
				if (value <= 0u) throw new ArgumentOutOfRangeException("value");
				_maximumSolutionsToFind = value; 
			}
		}

		/// <summary>Gets or sets the techniques used for solving.</summary>
		public List<EliminationTechnique> EliminationTechniques 
		{ 
			get 
			{
				if (_techniques == null) _techniques = new List<EliminationTechnique>();
				return _techniques; 
			} 
			set { _techniques = value; }
		}

		/// <summary>Gets or sets whether to allow brute-force techniques in the solver.</summary>
		public bool AllowBruteForce { get { return _allowBruteForce; } set { _allowBruteForce = value; } }

		/// <summary>Clones the solver options.</summary>
		/// <returns>A clone of the options.</returns>
		public SolverOptions Clone()
		{
            SolverOptions options = new SolverOptions();
            options.AllowBruteForce = AllowBruteForce;
            options.MaximumSolutionsToFind = MaximumSolutionsToFind;
            options.EliminationTechniques = new List<EliminationTechnique>();
            foreach (EliminationTechnique technique in EliminationTechniques)
            {
                options.EliminationTechniques.Add(technique.Clone());
            }
            return options;
		}

		/// <summary>Clones the solver options.</summary>
		/// <returns>A clone of the options.</returns>
		object ICloneable.Clone() { return Clone(); }
	}
}